src/Traversals/depthfirst.jl

```diff
@@ -95,14 +95,13 @@ end
     vcolor = zeros(UInt8, nv(g))
     verts = Vector{T}()
     state = TopoSortState(vcolor, verts, zero(T))
-    for v in vertices(g)
-        state.vcolor[v] != 0 && continue
-        state.vcolor[v] = 1
-        if !traverse_graph!(g, v, DepthFirst(), state)
-            throw(CycleError())
-        end
-    end
-    return reverse(state.verts)
+    
+    sources = filter(x -> indegree(g, x) == 0, vertices(g))
+ 
+    traverse_graph!(g, sources, DepthFirst(), state) || throw(CycleError())
+      
+    length(state.verts) == nv(g) || throw(CycleError())
```

Comment:
Let's be more specific here. If there's a cycle, will the length be greater than or less than `nv(g)`?

Reply:
if there is a cycle that can’t be reached from any of the sources, it won’t be in state.verts, so if
 `length(state.verts) < nv(g)` that means there is a cycle

Reply:
ok, let's make that explicit, and if the length is greater than throw a `TraversalError` of some sort.

Reply:
won’t that make coverage issue if there is no test that will throws that error, can i make assert ?

Reply:
That's the beauty of this: a short-circuiting statement gets evaluated so there's no coverage issue. That is,

```
true || throw(TraversalError())
```
will not fail code coverage. :)